File: services/ingest/frontend/src/lib/webhooks.ts
/**
 * Types + tiny helpers for the webhook delivery surface.
 * Field names mirror backend/app/schemas/webhooks.py exactly.
 */

import { ApiError, apiFetch, apiJson } from '@/lib/api';

/** Exact event strings the backend emits -- see webhook_connections.events. */
export type WebhookEvent =
  | 'job.finished'
  | 'job.failed'
  | 'import_run.finished'
  | 'document.processed'
  | 'collection.updated';

export const WEBHOOK_EVENTS: WebhookEvent[] = [
  'job.finished',
  'job.failed',
  'import_run.finished',
  'document.processed',
  'collection.updated',
];

/** Plain-language labels for the events checkbox group. */
export const webhookEventLabel: Record<WebhookEvent, string> = {
  'job.finished': 'Job finished',
  'job.failed': 'Job failed',
  'import_run.finished': 'Import run finished',
  // See contracts/events/document.processed.md -- dispatched alongside
  // job.finished for a successful job, carrying frontmatter + quality-gate
  // data plus a markdown_url instead of the inline markdown job.finished sends.
  'document.processed': 'Document processed',
  // See contracts/events/collection.updated.md -- fired on POST /collections
  // and PATCH /collections/{id} (not a job/run completion), fanned out to
  // every subscribed connection rather than a single per-task one.
  'collection.updated': 'Collection updated',
};

export type WebhookConnection = {
  id: string;
  name: string;
  url: string;
  /** Never the secret itself -- just whether one is on file. */
  has_secret: boolean;
  enabled: boolean;
  events: WebhookEvent[];
  created_at: string;
  updated_at: string;
};

export type WebhookConnectionListResponse = {
  items: WebhookConnection[];
};

export type WebhookConnectionCreateRequest = {
  name: string;
  url: string;
  secret?: string;
  enabled: boolean;
  events: WebhookEvent[];
};

/** PATCH body -- secret omitted or empty keeps the stored secret. */
export type WebhookConnectionUpdateRequest = {
  name?: string;
  url?: string;
  secret?: string;
  enabled?: boolean;
  events?: WebhookEvent[];
};

export type WebhookConnectionTestResponse = {
  ok: boolean;
  detail: string | null;
};

export type WebhookDeliveryStatus = 'pending' | 'sent' | 'failed';

export type WebhookDelivery = {
  id: string;
  connection_id: string | null;
  connection_name: string;
  event: WebhookEvent;
  job_id: string | null;
  import_run_id: string | null;
  collection_id: string | null;
  status: WebhookDeliveryStatus;
  http_status: number | null;
  error_message: string | null;
  attempts: number;
  created_at: string;
  updated_at: string;
};

export type WebhookDeliveryListResponse = {
  items: WebhookDelivery[];
};

export type WebhookSendRequest = {
  connection_id: string;
  job_id: string;
};

/** Status chip classes, matching the app's existing status-badge visual language. */
export const webhookDeliveryStatusChip: Record<WebhookDeliveryStatus, string> = {
  pending: 'bg-slate-100 text-slate-700',
  sent: 'bg-emerald-100 text-emerald-800',
  failed: 'bg-red-100 text-red-700',
};

/**
 * Matches the openwebui_test_cooldown_seconds-style per-connection test
 * cooldown documented for POST /webhooks/connections/{id}/test. The 429
 * does carry a Retry-After header, but apiJson's ApiError only exposes the
 * parsed detail string, not response headers -- so the client falls back to
 * this fixed window, same as OPENWEBUI_TEST_COOLDOWN_FALLBACK_MS (lib/openwebui.ts).
 */
export const WEBHOOK_TEST_COOLDOWN_FALLBACK_MS = 10_000;

const BASE = '/api/v1/webhooks';

/**
 * Like apiFetch + ok-check, but for the DELETE endpoint whose success body
 * we do not need. Duplicates lib/openwebui.ts's `send` rather than
 * importing it -- lib/ files stay independent of each other in this
 * codebase.
 */
async function send(path: string, init?: RequestInit): Promise<void> {
  const res = await apiFetch(path, init);
  if (!res.ok) {
    let detail = `Request failed with status ${res.status}`;
    try {
      const body = await res.json();
      if (typeof body?.detail === 'string') detail = body.detail;
    } catch {
      // Non-JSON error body -- keep the generic message.
    }
    throw new ApiError(res.status, detail);
  }
}

export function listWebhookConnections(): Promise<WebhookConnectionListResponse> {
  return apiJson<WebhookConnectionListResponse>(`${BASE}/connections`, { cache: 'no-store' });
}

export function createWebhookConnection(payload: WebhookConnectionCreateRequest): Promise<WebhookConnection> {
  return apiJson<WebhookConnection>(`${BASE}/connections`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  });
}

export function updateWebhookConnection(
  id: string,
  payload: WebhookConnectionUpdateRequest,
): Promise<WebhookConnection> {
  return apiJson<WebhookConnection>(`${BASE}/connections/${id}`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  });
}

export function deleteWebhookConnection(id: string): Promise<void> {
  return send(`${BASE}/connections/${id}`, { method: 'DELETE' });
}

export function testWebhookConnection(id: string): Promise<WebhookConnectionTestResponse> {
  return apiJson<WebhookConnectionTestResponse>(`${BASE}/connections/${id}/test`, { method: 'POST' });
}

/** GET /webhooks/deliveries -- the caller's own recent deliveries, newest first. */
export function listWebhookDeliveries(params: { limit?: number } = {}): Promise<WebhookDeliveryListResponse> {
  const query = new URLSearchParams();
  if (params.limit) query.set('limit', String(params.limit));
  const qs = query.toString();
  return apiJson<WebhookDeliveryListResponse>(`${BASE}/deliveries${qs ? `?${qs}` : ''}`, { cache: 'no-store' });
}

/** POST /webhooks/send -- manual delivery of one FINISHED job to one connection. */
export function sendWebhook(payload: WebhookSendRequest): Promise<WebhookDelivery> {
  return apiJson<WebhookDelivery>(`${BASE}/send`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  });
}

File: services/ingest/backend/app/schemas/webhooks.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

# Fixed event set (exact strings) -- also the payload contract's `event`
# field values. Kept here as the single source of truth: request schemas
# validate against it via the Literal type below (invalid entries 422 for
# free), and app/api/webhook_routes.py imports this tuple for the same
# purpose wherever a plain Python container is more convenient than a type.
#
# 'document.processed' (see contracts/events/document.processed.md) rides
# the exact same connection/delivery/retry machinery as job.finished --
# it is dispatched right alongside job.finished (never on its own, never for
# job.failed) by app/workers/tasks.py's completion hook, through
# app/workers/webhook_tasks.dispatch_job_event with a second event string.
WEBHOOK_EVENTS: tuple[str, ...] = (
    'job.finished', 'job.failed', 'import_run.finished', 'document.processed',
)

WebhookEvent = Literal[
    'job.finished', 'job.failed', 'import_run.finished', 'document.processed',
]


# --- Connections ------------------------------------------------------------

class WebhookConnectionCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    # Preserved verbatim after validation because paths and query strings can
    # be meaningful to generic export receivers.
    url: str = Field(min_length=1, max_length=2048)
    # Write-only and optional: a connection may be created without a signing
    # secret, though signed exports are recommended.
    secret: str | None = Field(default=None, max_length=4096)
    events: list[WebhookEvent] = Field(min_length=1, max_length=len(WEBHOOK_EVENTS))
    enabled: bool = True


class WebhookConnectionUpdateRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    url: str | None = Field(default=None, min_length=1, max_length=2048)
    events: list[WebhookEvent] | None = Field(default=None, min_length=1, max_length=len(WEBHOOK_EVENTS))
    enabled: bool | None = None
    # Write-only update, tri-state via model_fields_set (see
    # app/api/webhook_routes.update_webhook_connection):
    #   - key omitted entirely            -> stored secret unchanged
    #   - key present, value None or ''   -> stored secret cleared (secret_encrypted = NULL)
    #   - key present, non-empty value    -> stored secret rotated
    # The secret column is nullable: omitted means "keep", while an explicit
    # empty value clears it.
    secret: str | None = Field(default=None, max_length=4096)


class WebhookConnectionResponse(BaseModel):
    id: str
    name: str
    url: str
    enabled: bool
    events: list[str]
    # Never the secret itself -- just whether one is on file.
    has_secret: bool = True
    created_at: datetime
    updated_at: datetime

    model_config = {'from_attributes': True}


class WebhookConnectionListResponse(BaseModel):
    items: list[WebhookConnectionResponse]


class WebhookConnectionTestResponse(BaseModel):
    ok: bool
    detail: str | None = None
    http_status: int | None = None


# --- Deliveries ---------------------------------------------------------------

class WebhookSendRequest(BaseModel):
    connection_id: str = Field(min_length=1)
    job_id: str = Field(min_length=1)


class WebhookDeliveryResponse(BaseModel):
    id: str
    connection_id: str | None = None
    connection_name: str
    event: str
    job_id: str | None = None
    import_run_id: str | None = None
    collection_id: str | None = None
    status: Literal['pending', 'sent', 'failed']
    http_status: int | None = None
    error_message: str | None = None
    attempts: int
    created_at: datetime
    updated_at: datetime

    model_config = {'from_attributes': True}


class WebhookDeliveryListResponse(BaseModel):
    items: list[WebhookDeliveryResponse]
